### PR TITLE
Backward ops for Metalium - part-2

### PR DESCRIPTION
**Pull request by muthutt**
week 3, 4, backward ops sprint include following ops support,
#4285: Add backward support for sub (unary and binary)
#4306: MAE, MSE loss functions
#4306: Update sum reduction mode with loop
#4284: Add backward support for min, max op
#4283: Add backward support for binary assign op
#4331: Add  backward ops for rsub, and log ops
#4338: Add support for backward neg op
#4203: Implementation of Lamb optimizer
#4339: Add support for rsqrt backward op
#4288: Add backward op support for abs
#4341: Add backward support for LT, GT and NE op

### CODE REVIEW

**Review by eyonland — APPROVED** [review]

File: tests/tt_eager/python_api_testing/unit_testing/backward_ops/test_backward_binary_assign.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    in_data.retain_grad()
+
+    pyt_y = torch.clone(in_data)
+
+    pyt_y.backward(gradient=grad_data)
+
+    golden_output_tensor = in_data.grad
+
+    comp_pass, _ = comparison_funcs.comp_equal(golden_output_tensor, tt_output_tensor)
+    _, comp_out = comparison_funcs.comp_allclose_and_pcc(golden_output_tensor, tt_output_tensor)
```
Comment (tt-aho) [inline tests/tt_eager/python_api_testing/unit_testing/backward_ops/test_backward_binary_assign.py:52]: comp_equal returns the atol, rtol, pcc in it's output, so you dont need to run comp_allclose_and_pcc separately to get comp_out

File: tests/tt_eager/python_api_testing/unit_testing/backward_ops/test_backward_embedding.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        input_index.reshape((batch_size, no_of_embeddings)),
+        weights.reshape((batch_size * no_of_embeddings, embedding_dim)),
+    ).reshape((1, 1, batch_size * no_of_embeddings, embedding_dim))
+
+    pyt_y.backward(gradient=grad_data)
+
+    golden_output_tensor_a = weights.grad
+
+    comp_pass_a, _ = comparison_funcs.comp_pcc(golden_output_tensor_a, tt_output_tensor_a, 0.99)
+    _, comp_out_a = comparison_funcs.comp_allclose_and_pcc(golden_output_tensor_a, tt_output_tensor_a)
```
Comment (tt-aho) [inline tests/tt_eager/python_api_testing/unit_testing/backward_ops/test_backward_embedding.py:61]: Same as above comp_pcc reports all the necessary info

File: tests/tt_eager/python_api_testing/unit_testing/loss_ops/test_loss_mae.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+        tt_output_tensor_on_device = tt_lib.tensor.maeloss(
+            ref_tensor, pred_tensor, tt_lib.tensor.LossReductionMode.NONE, memcfg
+        )
+        tt_mae_output = tt_output_tensor_on_device.cpu().to(tt_lib.tensor.Layout.ROW_MAJOR).to_torch()
+
+        loss = torch.nn.L1Loss(reduction="none")
+        pt_mae_output = loss(ref_data.to(torch.float32), pred_data.to(torch.float32))
+        comp_pass_a, _ = comparison_funcs.comp_pcc(pt_mae_output, tt_mae_output)
+        _, comp_out_a = comparison_funcs.comp_allclose_and_pcc(pt_mae_output, tt_mae_output)
```
Comment (tt-aho) [inline tests/tt_eager/python_api_testing/unit_testing/loss_ops/test_loss_mae.py:51]: Same as above

File: tests/tt_eager/python_api_testing/unit_testing/loss_ops/test_loss_mse.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+        tt_output_tensor_on_device = tt_lib.tensor.mseloss(
+            ref_tensor, pred_tensor, tt_lib.tensor.LossReductionMode.NONE, memcfg
+        )
+        tt_mse_output = tt_output_tensor_on_device.cpu().to(tt_lib.tensor.Layout.ROW_MAJOR).to_torch()
+
+        loss = torch.nn.MSELoss(reduction="none")
+        pt_mse_output = loss(ref_data.to(torch.float32), pred_data.to(torch.float32))
+        comp_pass_a, _ = comparison_funcs.comp_pcc(pt_mse_output, tt_mse_output)
+        _, comp_out_a = comparison_funcs.comp_allclose_and_pcc(pt_mse_output, tt_mse_output)
```
Comment (tt-aho) [inline tests/tt_eager/python_api_testing/unit_testing/loss_ops/test_loss_mse.py:51]: Same as above

File: tt_eager/tt_lib/csrc/tt_lib_bindings_tensor.cpp [hunk 5]
```diff
@@ -100,6 +101,10 @@ void TensorModule(py::module &m_tensor) {
 
     detail::export_enum<ShardOrientation>(m_tensor);
 
+    detail::export_enum<LossFunction>(m_tensor);
+
```
Comment (tt-aho) [inline tt_eager/tt_lib/csrc/tt_lib_bindings_tensor.cpp:105]: Any reason to export this to the user if we aren't exposing the function that uses this?

**Review by tt-aho — CHANGES_REQUESTED** [review]

**Review by mywoodstock — APPROVED** [review]

File: tests/tt_eager/python_api_testing/unit_testing/backward_ops/test_backward_binary_assign.py (hunk 1 above)
Comment (umadevimcw) [reply tests/tt_eager/python_api_testing/unit_testing/backward_ops/test_backward_binary_assign.py]: Done

File: tests/tt_eager/python_api_testing/unit_testing/backward_ops/test_backward_sigmoid.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    tt_output_tensor = tt_output_tensor_on_device[0].cpu().to(tt_lib.tensor.Layout.ROW_MAJOR).to_torch()
+
+    in_data.retain_grad()
+
+    pyt_y.backward(gradient=grad_data)
+
+    golden_output_tensor = in_data.grad
+
+    comp_pass, _ = comparison_funcs.comp_pcc(golden_output_tensor, tt_output_tensor, 0.90)
+    _, comp_out = comparison_funcs.comp_allclose_and_pcc(golden_output_tensor, tt_output_tensor)
```
[the remark this answers is not in the document]
Comment (umadevimcw) [reply tests/tt_eager/python_api_testing/unit_testing/backward_ops/test_backward_sigmoid.py]: @tt-aho  Done

File: tests/tt_eager/python_api_testing/unit_testing/loss_ops/test_loss_mse.py (hunk 4 above)
Comment (umadevimcw) [reply tests/tt_eager/python_api_testing/unit_testing/loss_ops/test_loss_mse.py:51]: @tt-aho @muthutt  Done

File: tests/tt_eager/python_api_testing/unit_testing/backward_ops/test_backward_embedding.py (hunk 2 above)
Comment (umadevimcw) [reply tests/tt_eager/python_api_testing/unit_testing/backward_ops/test_backward_embedding.py]: @tt-aho Done

**Review by tt-aho — APPROVED** [review]

**Review by TT-billteng — CHANGES_REQUESTED** [review]
This change shouldn't be changing metal or bumping umd submodule

**Review by TT-billteng — APPROVED** [review]
